app/controllers/tickets_controller.rb

```diff
@@ -173,11 +175,13 @@ def new
 
   def create
     # the hook that is triggered when receiving an email.
-    if params[:format] == 'json'
+    if params[:hook]
+      unless Devise.secure_compare(params[:mail_key], MAIL_KEY)
+        render status: :forbidden, text: t(:access_denied)
+        return
+      end
       using_hook = true # we assume different policies to create a ticket when we receive an email
-      base64_message = ((params[:base64] == true) || !(params[:message][0,64] =~ /^([A-Za-z0-9+\/]{4})*([A-Za-z0-9+\/]{4}|[A-Za-z0-9+\/]{3}=|[A-Za-z0-9+\/]{2}==)$/).nil?)
-      message = base64_message ? Base64.decode64(params[:message].strip) : params[:message]
-      @ticket = TicketMailer.receive(message)
+      @ticket = TicketMailer.receive(send(params[:hook]))
```

Comment:
You're letting "user" input call a function here. I saw your validation in the routes, but please add some test to make sure we cannot call other functions, to avoid problems in the future, when somebody changes that.

Reply:
Unfortunately, the exclusiveness of a constraint Regexp is not something you can cover by any test. But I see your concern and I've fortified this in three ways now:

1. I've prefixed the hook handler methods with `raw_` (e.g. `raw_mailgun`). It's therefore impossible to call arbitrary methods even if the routes are tainted. 
2. The list of acceptable hooks resides in `TicketsController::MAIL_HOOKS` now, the constraint is no longer a Regexp but a Lambda which does checks for inclusion.
3. As per your request, I've added a simple test to check the contents of `TicketsController::MAIL_HOOKS`. This failsafe is a bit of an overkill (given 2 above), but better safe than sorry.